File: lib/scripts/mouse.py
import json
import sys
import pyautogui as auto


def map_tween(tween):
    if tween == "linear":
        return auto.linear
    elif tween == "ease-in":
        return auto.easeInQuad
    elif tween == "ease-out":
        return auto.easeOutQuad
    elif tween == "ease-in-out":
        return auto.easeInOutQuad
    elif tween == 'bounce':
        return auto.easeInBounce
    elif tween == "elastic":
        return auto.easeInElastic
    else:
        return auto.linear


class Mouse:

    @staticmethod
    def get_mouse_pos() -> tuple:
        x, y = auto.position()
        data = {
            'x': x,
            'y': y
        }
        return json.dumps(data)

    @staticmethod
    def move_to(x: int = 0, y: int = 0, duration: int = 0, tween: str = "linear") -> None:
        tween_func = map_tween(tween)
        auto.moveTo(x=x, y=y, duration=duration, tween=tween_func)

    @staticmethod
    def move_rel(x: int = 0, y: int = 0, duration: int = 0) -> None:
        auto.moveRel(xOffset=x, yOffset=y, duration=duration)

    @staticmethod
    def drag_to(x: int = 0, y: int = 0, duration: int = 0, tween: str = "linear", button: str = "left") -> None:
        tween_func = map_tween(tween)
        auto.dragTo(x, y, duration, tween_func, button)

    @staticmethod
    def drag_rel(x: int = 0, y: int = 0, duration: int = 0, tween: str = "linear", button: str = "left") -> None:
        tween_func = map_tween(tween)
        auto.dragTo(x, y, duration, tween_func, button)

    @staticmethod
    def click(x: int = 0, y: int = 0, clicks: int = 1, interval: int = 0, button: str = "left", duration: int = 0, tween: str = "linear"):
        tween_func = map_tween(tween)
        auto.click(x, y, clicks, interval, button, duration, tween_func)

    @staticmethod
    def default(): return json.dumps({})


if __name__ == "__main__":
    arg = sys.argv
    command = arg[1]

    if command == "pos":
        print(Mouse.get_mouse_pos())
    elif command == "move_to":
        x, y, dur, tween = arg[2], arg[3], arg[4], arg[5]
        Mouse.move_to(int(x), int(y), int(dur), str(tween))
    elif command == "move_rel":
        x, y, dur = arg[2], arg[3], arg[4]
        Mouse.move_rel(int(x), int(y), int(dur))
    elif command == "drag_to":
        x, y, dur, tween, button = arg[2], arg[3], arg[4], arg[5], arg[6]
        Mouse.drag_to(int(x), int(y), int(dur), str(tween), str(button))
    elif command == "drag_rel":
        x, y, dur, tween, button = arg[2], arg[3], arg[4], arg[5], arg[6]
        Mouse.drag_rel(int(x), int(y), int(dur), str(tween), str(button))
    elif command == "click":
        x, y, clicks, interval, button, duration, tween = arg[
            2], arg[3], arg[4], arg[5], arg[6], arg[7], arg[8]
        Mouse.click(int(x), int(y), int(clicks), int(interval),
                    str(button), int(duration), str(tween))

    else:
        print(Mouse.default())

File: lib/src/mouse.dart
import 'dart:convert';
import 'dart:io';

import 'package:dart_autogui/src/utils.dart';

/// `MousePosition` holds the data for where mouse is
/// returns `x` and `y` positions on screen
class MousePosition {
  final int x;
  final int y;

  MousePosition({
    required this.x,
    required this.y,
  });
}

const _kDefaultTween = "linear";
const _kDefaultButton = "left";

/// Controller for the mouse tweening
/// By default `MouseTween.linear` is used in twwn functons
enum MouseTween {
  linear,
  easeInQuad,
  easeOutQuad,
  easeInOutQuad,
  easeInBounce,
  easeInElastic
}

/// Controller for Mouse button
enum MouseButton {
  left,
  middle,
  right,
}

/// Internal function that maps tween function into String
String _mapTween(MouseTween? tween) {
  if (tween != null) {
    if (tween == MouseTween.linear) {
      return _kDefaultTween;
    } else if (tween == MouseTween.easeInQuad) {
      return 'ease-in';
    } else if (tween == MouseTween.easeOutQuad) {
      return 'ease-out';
    } else if (tween == MouseTween.easeInOutQuad) {
      return 'ease-in-out';
    } else if (tween == MouseTween.easeInBounce) {
      return 'bounce';
    } else if (tween == MouseTween.easeInElastic) {
      return 'elastic';
    } else {
      return _kDefaultTween;
    }
  } else {
    return _kDefaultTween;
  }
}

/// Internal function that maps button controller into String
String _mapButton(MouseButton? button) {
  if (button != null) {
    if (button == MouseButton.left) {
      return _kDefaultButton;
    } else if (button == MouseButton.right) {
      return 'right';
    } else {
      return _kDefaultButton;
    }
  } else {
    return _kDefaultButton;
  }
}

/// Main `Mouse` object
class Mouse {
  /// Get mouse position ,
  /// returns `Future<MousePosition>`
  static Future<MousePosition> pos() async {
    final path = getScriptPath('mouse.py');
    final data = await Process.run('python3', [path, 'pos']);
    final pos = jsonDecode(data.stdout.toString());

    return MousePosition(x: pos['x'], y: pos['y']);
  }

  /// Move mouse to a certain position ,
  /// returns `Future<void>`
  static Future<void> moveTo({
    required int x,
    required int y,
    Duration? duration,
    MouseTween? tween,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'move_to',
      x.toString(),
      y.toString(),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
    ]);
  }

  /// Move mouse cursor relative to the current mouse position ,
  /// returns `Future<void>`
  static Future<void> moveRel({
    required int x,
    required int y,
    Duration? duration,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'move_rel',
      x.toString(),
      y.toString(),
      duration != null ? duration.inSeconds.toString() : '0',
    ]);
  }

  /// drags mouse cursor to a certain position ,
  /// returns `Future<void>`
  static Future<void> dragTo({
    required int x,
    required int y,
    Duration? duration,
    MouseTween? tween,
    MouseButton? button,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'drag_to',
      x.toString(),
      y.toString(),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
      _mapButton(button),
    ]);
  }

  /// Drag mouse cursor relative to the current mouse position ,
  /// returns `Future<void>`
  static Future<void> dragRel({
    required int x,
    required int y,
    Duration? duration,
    MouseTween? tween,
    MouseButton? button,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'drag_rel',
      x.toString(),
      y.toString(),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
      _mapButton(button),
    ]);
  }

  /// Clicks on certain position
  /// returns `Future<void>`
  static Future<void> click({
    required int x,
    required int y,
    int clicks = 1,
    int intervals = 0,
    Duration? duration,
    MouseTween? tween,
    MouseButton? button,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'click',
      x.toString(),
      y.toString(),
      clicks.toString(),
      intervals.toString(),
      _mapButton(button),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
    ]);
  }

  /// Right clicks on certain position
  /// returns `Future<void>`
  static Future<void> rightClick({
    required int x,
    required int y,
    int clicks = 1,
    int intervals = 0,
    Duration? duration,
    MouseTween? tween,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'click',
      x.toString(),
      y.toString(),
      clicks.toString(),
      intervals.toString(),
      _mapButton(MouseButton.right),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
    ]);
  }

  /// Left clicks on certain position
  /// returns `Future<void>`
  static Future<void> leftClick({
    required int x,
    required int y,
    int clicks = 1,
    int intervals = 0,
    Duration? duration,
    MouseTween? tween,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'click',
      x.toString(),
      y.toString(),
      clicks.toString(),
      intervals.toString(),
      _mapButton(MouseButton.left),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
    ]);
  }

  /// Middle clicks on certain position
  /// returns `Future<void>`
  static Future<void> middleClick({
    required int x,
    required int y,
    int clicks = 1,
    int intervals = 0,
    Duration? duration,
    MouseTween? tween,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'click',
      x.toString(),
      y.toString(),
      clicks.toString(),
      intervals.toString(),
      _mapButton(MouseButton.middle),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
    ]);
  }

  /// Double clicks on certain position (in right button) ,
  /// returns `Future<void>`
  static Future<void> doubleClick({
    required int x,
    required int y,
    int intervals = 0,
    Duration? duration,
    MouseTween? tween,
    MouseButton? button,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'click',
      x.toString(),
      y.toString(),
      2.toString(),
      intervals.toString(),
      _mapButton(button),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
    ]);
  }

  /// Triple clicks on certain position (in right button) ,
  /// returns `Future<void>`
  static Future<void> trippleClick({
    required int x,
    required int y,
    int intervals = 0,
    Duration? duration,
    MouseTween? tween,
    MouseButton? button,
  }) async {
    final path = getScriptPath('mouse.py');
    await Process.run('python3', [
      path,
      'click',
      x.toString(),
      y.toString(),
      3.toString(),
      intervals.toString(),
      _mapButton(button),
      duration != null ? duration.inSeconds.toString() : '0',
      _mapTween(tween),
    ]);
  }
}
